lib/mayaUsd/render/vp2ShaderFragments/shaderFragments.cpp

```diff
@@ -375,28 +389,30 @@ MStatus HdVP2ShaderFragments::registerFragments()
         //    Where the ... denotes a 4x4 matrix expanded from the 3x3 matrix passed in as parameter
         //    to the function (alpha is always left untouched).
         //
-        auto registerSpace = [&](const std::string& spaceName,
+        auto registerSpace = [&](const std::string& ocioName,
+                                 const std::string& synColorName,
                                  const std::string& fragName,
                                  const float*       convMatrix) {
             if (!fragmentManager->hasFragment(fragName.c_str())) {
                 std::string xmlFinal = std::regex_replace(xmlString, RE_FRAG, fragName.c_str());
                 if (convMatrix) {
-                    std::stringstream op_glsl, op_hlsl, op_cg;
+                    std::stringstream op_glsl, op_hlsl, op_cg, op_matrix;
                     op_glsl << "outColor = mat4(";
                     op_hlsl << "outColor = mul(outColor, float4x4(";
+                    op_matrix << convMatrix[0];
+                    for (int i = 1; i < 16; ++i) {
+                        op_matrix << ", " << convMatrix[i];
+                    }
+                    op_glsl << op_matrix.str() << ") * outColor;";
+                    op_hlsl << op_matrix.str() << "));";
                     op_cg << "outColor = mul(float4x4(";
-                    for (int i = 0; i < 3; ++i) {
-                        op_glsl << convMatrix[3 * i] << ", " << convMatrix[3 * i + 1] << ", "
-                                << convMatrix[3 * i + 2] << ", " << 0.0 << ", ";
-                        op_hlsl << convMatrix[3 * i] << ", " << convMatrix[3 * i + 1] << ", "
-                                << convMatrix[3 * i + 2] << ", " << 0.0 << ", ";
+                    for (int i = 0; i < 4; ++i) {
                         // Cg needs transposed matrix:
-                        op_cg << convMatrix[i] << ", " << convMatrix[3 + i] << ", "
-                              << convMatrix[6 + i] << ", " << 0.0 << ", ";
+                        op_cg << convMatrix[i] << ", " << convMatrix[4 + i] << ", "
+                              << convMatrix[8 + i] << ", " << convMatrix[12 + i]
+                              << ((i < 3) ? ", " : "");
                     }
-                    op_glsl << 0.0 << ", " << 0.0 << ", " << 0.0 << ", " << 1.0 << ") * outColor;";
-                    op_hlsl << 0.0 << ", " << 0.0 << ", " << 0.0 << ", " << 1.0 << "));";
-                    op_cg << 0.0 << ", " << 0.0 << ", " << 0.0 << ", " << 1.0 << "), outColor);";
+                    op_cg << "), outColor);";
```

Comment:
I still have investigations to perform about the matrix layout. But you can use the OpenColorIO implementation to at least have the same 'result' than OpenColorIO v2.
Refer to https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/master/src/OpenColorIO/GpuShaderUtils.cpp#L671

Reply:
The result is the same because the color transformation matrices used here are already pre-transposed to Dx11/OpenGL order. 

This allows making Cg the odd one out requiring back-transposition, which is OK since that language was deprecated nine years ago.

Reply:
> The result is the same because the color transformation matrices used here are already pre-transposed to Dx11/OpenGL order.

Could you add a comment to mention that matrices are already transposed because that's really not clear in the code?

Reply:
Done.